File: visitor/typescript/Visitor.ts
// Visitor interface
interface ShapeVisitor {
  visitCircle(circle: Circle): string;
  visitRectangle(rectangle: Rectangle): string;
  visitTriangle(triangle: Triangle): string;
}

// Element interface
interface Shape {
  accept(visitor: ShapeVisitor): string;
}

// Concrete Elements
class Circle implements Shape {
  constructor(public radius: number) {}

  accept(visitor: ShapeVisitor): string {
    return visitor.visitCircle(this);
  }
}

class Rectangle implements Shape {
  constructor(public width: number, public height: number) {}

  accept(visitor: ShapeVisitor): string {
    return visitor.visitRectangle(this);
  }
}

class Triangle implements Shape {
  constructor(public base: number, public height: number) {}

  accept(visitor: ShapeVisitor): string {
    return visitor.visitTriangle(this);
  }
}

// Concrete Visitors
class AreaCalculator implements ShapeVisitor {
  visitCircle(circle: Circle): string {
    const area = Math.PI * circle.radius * circle.radius;
    return `Circle area: ${area.toFixed(2)}`;
  }

  visitRectangle(rectangle: Rectangle): string {
    const area = rectangle.width * rectangle.height;
    return `Rectangle area: ${area.toFixed(2)}`;
  }

  visitTriangle(triangle: Triangle): string {
    const area = 0.5 * triangle.base * triangle.height;
    return `Triangle area: ${area.toFixed(2)}`;
  }
}

class DrawVisitor implements ShapeVisitor {
  visitCircle(circle: Circle): string {
    return `Drawing Circle with radius ${circle.radius.toFixed(2)}`;
  }

  visitRectangle(rectangle: Rectangle): string {
    return `Drawing Rectangle ${rectangle.width.toFixed(2)} x ${rectangle.height.toFixed(2)}`;
  }

  visitTriangle(triangle: Triangle): string {
    return `Drawing Triangle base ${triangle.base.toFixed(2)}, height ${triangle.height.toFixed(2)}`;
  }
}

// Usage
const shapes: Shape[] = [
  new Circle(5),
  new Rectangle(4, 6),
  new Triangle(3, 4)
];

const areaCalc = new AreaCalculator();
const drawer = new DrawVisitor();

console.log('Calculating areas:');
shapes.forEach(shape => console.log(shape.accept(areaCalc)));

console.log('\nDrawing shapes:');
shapes.forEach(shape => console.log(shape.accept(drawer)));

File: visitor/java/Visitor.java
// Visitor interface
interface ShapeVisitor {
    String visitCircle(Circle circle);
    String visitRectangle(Rectangle rectangle);
    String visitTriangle(Triangle triangle);
}

// Element interface
interface Shape {
    String accept(ShapeVisitor visitor);
}

// Concrete Elements
class Circle implements Shape {
    private double radius;

    public Circle(double radius) {
        this.radius = radius;
    }

    public double getRadius() {
        return radius;
    }

    @Override
    public String accept(ShapeVisitor visitor) {
        return visitor.visitCircle(this);
    }
}

class Rectangle implements Shape {
    private double width;
    private double height;

    public Rectangle(double width, double height) {
        this.width = width;
        this.height = height;
    }

    public double getWidth() {
        return width;
    }

    public double getHeight() {
        return height;
    }

    @Override
    public String accept(ShapeVisitor visitor) {
        return visitor.visitRectangle(this);
    }
}

class Triangle implements Shape {
    private double base;
    private double height;

    public Triangle(double base, double height) {
        this.base = base;
        this.height = height;
    }

    public double getBase() {
        return base;
    }

    public double getHeight() {
        return height;
    }

    @Override
    public String accept(ShapeVisitor visitor) {
        return visitor.visitTriangle(this);
    }
}

// Concrete Visitors
class AreaCalculator implements ShapeVisitor {
    @Override
    public String visitCircle(Circle circle) {
        double area = Math.PI * circle.getRadius() * circle.getRadius();
        return String.format("Circle area: %.2f", area);
    }

    @Override
    public String visitRectangle(Rectangle rectangle) {
        double area = rectangle.getWidth() * rectangle.getHeight();
        return String.format("Rectangle area: %.2f", area);
    }

    @Override
    public String visitTriangle(Triangle triangle) {
        double area = 0.5 * triangle.getBase() * triangle.getHeight();
        return String.format("Triangle area: %.2f", area);
    }
}

class DrawVisitor implements ShapeVisitor {
    @Override
    public String visitCircle(Circle circle) {
        return String.format("Drawing Circle with radius %.2f", circle.getRadius());
    }

    @Override
    public String visitRectangle(Rectangle rectangle) {
        return String.format("Drawing Rectangle %.2f x %.2f", rectangle.getWidth(), rectangle.getHeight());
    }

    @Override
    public String visitTriangle(Triangle triangle) {
        return String.format("Drawing Triangle base %.2f, height %.2f", triangle.getBase(), triangle.getHeight());
    }
}

// Main class demonstrating Visitor pattern
public class Visitor {
    public static void main(String[] args) {
        Shape[] shapes = {
            new Circle(5),
            new Rectangle(4, 6),
            new Triangle(3, 4)
        };

        ShapeVisitor areaCalc = new AreaCalculator();
        ShapeVisitor drawer = new DrawVisitor();

        System.out.println("Calculating areas:");
        for (Shape shape : shapes) {
            System.out.println(shape.accept(areaCalc));
        }

        System.out.println("\nDrawing shapes:");
        for (Shape shape : shapes) {
            System.out.println(shape.accept(drawer));
        }
    }
}
